File: scripts/agent.template.ts
import fs from "https://deno.land/std@0.140.0/node/fs.ts";
import * as path from "jsr:@std/path";

let SHEETBOT_BASEURL;
if (!Deno.env.has("SHEETBOT_BASEURL")) {
  SHEETBOT_BASEURL
  SHEETBOT_BASEURL = "${req.protocol}://${req.get('host')}";
  Deno.env.set("SHEETBOT_BASEURL", SHEETBOT_BASEURL);
} else {
  SHEETBOT_BASEURL = Deno.env.get("SHEETBOT_BASEURL");
}

async function checkForErrors(responsePromise) {
  const response = await responsePromise;
  if (response.status === 401) {
    throw new Error("Unauthenticated");
  } else if (response.status === 403) {
    throw new Error("Unauthorized");
  } else if (response.status === 500) {
    throw new Error("Internal server error");
  } else if (response.status !== 200) {
    throw new Error("Some error occurred");
  }
  return response;
}

let headers = {};
if (Deno.env.has("SHEETBOT_AUTH_USER") && Deno.env.has("SHEETBOT_AUTH_PASS")) {
  const authr = await fetch(SHEETBOT_BASEURL + "/login", {
    method: "POST",
    body: JSON.stringify({
      username: Deno.env.get("SHEETBOT_AUTH_USER"),
      password: Deno.env.get("SHEETBOT_AUTH_PASS")
    }),
    headers: {
      "Content-Type": "application/json",
    },
  });
  Deno.env.delete("SHEETBOT_AUTH_USER");
  Deno.env.delete("SHEETBOT_AUTH_PASS");
  if (authr.status === 200) {
    const authj = await authr.json();
    headers["Authorization"] = `Bearer ${authj.token}`;
  } else {
    throw new Error("Login failed");
  }
}

let localCapabilities = {};
try {
  const capabilitiesJSONPath = path.resolve("./.capabilities.json");
  if (fs.existsSync(capabilitiesJSONPath)) {
    const capabilitiesText = Deno.readTextFileSync(capabilitiesJSONPath);
    localCapabilities = Object.assign({}, localCapabilities, JSON.parse(capabilitiesText));
  }

  const capabilitiesPath = path.resolve("./.capabilities.dynamic.ts");
  if (fs.existsSync(capabilitiesPath)) {
    const { getCapabilities } = await import(path.toFileUrl(capabilitiesPath).toString());
    const dynamicCapabilities = await getCapabilities(localCapabilities);
    localCapabilities = Object.assign({}, localCapabilities, dynamicCapabilities);
  }

  const capabilitiesOverrideJSONPath = path.resolve("./.capabilities.override.json");
  if (fs.existsSync(capabilitiesOverrideJSONPath)) {
    const capabilitiesOverrideText = Deno.readTextFileSync(capabilitiesOverrideJSONPath);
    localCapabilities = Object.assign({}, localCapabilities, JSON.parse(capabilitiesOverrideText));
  }

  console.log("capabilities:", localCapabilities);
} catch (e) {
  console.log("Ignoring capabilities error:");
  console.log(e);
}

const capabilities = {
  os: Deno.build.os,
  arch: Deno.build.arch,
  ...localCapabilities
};

const response = await checkForErrors(fetch(SHEETBOT_BASEURL + "/tasks/get", {
  method: "POST",
  headers: {
    ...headers,
    "Content-Type": "application/json",
  },
  body: JSON.stringify({
    type: 'deno',
    capabilities: capabilities
  })
}));

const json = await response.json();
if (json.hasOwnProperty("script")) {
  Deno.env.set("SHEETBOT_TASK_ID", json.id);
  Deno.env.set("SHEETBOT_AUTHORIZATION_HEADER", headers["Authorization"]);
  Deno.env.set(
    "SHEETBOT_TASK_BASEURL",
    SHEETBOT_BASEURL + "/tasks/" + json.id,
  );
  Deno.env.set(
    "SHEETBOT_TASK_ACCEPTURL",
    SHEETBOT_BASEURL + "/tasks/" + json.id + "/accept",
  );
  Deno.env.set(
    "SHEETBOT_TASK_COMPLETEURL",
    SHEETBOT_BASEURL + "/tasks/" + json.id + "/complete",
  );
  Deno.env.set(
    "SHEETBOT_TASK_FAILEDURL",
    SHEETBOT_BASEURL + "/tasks/" + json.id + "/failed",
  );
  Deno.env.set(
    "SHEETBOT_TASK_DATAURL",
    SHEETBOT_BASEURL + "/tasks/" + json.id + "/data",
  );
  Deno.env.set(
    "SHEETBOT_TASK_ARTEFACTURL",
    SHEETBOT_BASEURL + "/tasks/" + json.id + "/artefacts",
  );

  await checkForErrors(fetch(Deno.env.get("SHEETBOT_TASK_ACCEPTURL")!, {
    method: "POST",
    headers: {
      ...headers,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({}),
  }));
  try {
    const data = await import(json.script);
    await checkForErrors(fetch(Deno.env.get("SHEETBOT_TASK_COMPLETEURL")!, {
      method: "POST",
      headers: {
        ...headers,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ data: data }),
    }));
  } catch (e) {
    console.error(e);
    await checkForErrors(fetch(Deno.env.get("SHEETBOT_TASK_FAILEDURL")!, {
      method: "POST",
      headers: {
        ...headers,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({}),
    }));
  }
}

File: scripts/agent.template.py
#!/usr/bin/env -S uv run --script
#
# /// script
# requires-python = ">=3.12"
# dependencies = ["requests"]
# ///
import os
import json
import requests
import sys
import platform
import importlib.util
import subprocess
import tempfile

if "SHEETBOT_BASEURL" not in os.environ:
    SHEETBOT_BASEURL = "${req.protocol}://${req.get('host')}"
    os.environ["SHEETBOT_BASEURL"] = SHEETBOT_BASEURL
else:
    SHEETBOT_BASEURL = os.environ["SHEETBOT_BASEURL"]


def check_for_errors(response):
    if response.status_code == 401:
        raise ValueError("Unauthenticated")
    elif response.status_code == 403:
        raise ValueError("Unauthorized")
    elif response.status_code == 500:
        raise ValueError("Internal server error")
    elif response.status_code != 200:
        raise ValueError("Some error occurred")
    return response


headers = {}
if os.environ.get("SHEETBOT_AUTH_USER") and os.environ.get("SHEETBOT_AUTH_PASS"):
    auth_response = requests.post(
        SHEETBOT_BASEURL + "/login",
        json={
            "username": os.environ["SHEETBOT_AUTH_USER"],
            "password": os.environ["SHEETBOT_AUTH_PASS"],
        },
    )
    del os.environ["SHEETBOT_AUTH_USER"]
    del os.environ["SHEETBOT_AUTH_PASS"]
    if auth_response.status_code == 200:
        auth_json = auth_response.json()
        headers["Authorization"] = f"Bearer {auth_json['token']}"
    else:
        raise ValueError("Login failed")

local_capabilities = {}
try:
    if os.path.exists("./.capabilities.json"):
        with open("./.capabilities.json") as f:
            local_capabilities.update(json.load(f))

    if os.path.exists("./.capabilities.dynamic.py"):
        spec = importlib.util.spec_from_file_location(
            "capabilities", "./.capabilities.dynamic.py"
        )
        if spec and spec.loader:
            module = importlib.util.module_from_spec(spec)
            spec.loader.exec_module(module)
            dynamic_capabilities = module.getCapabilities(local_capabilities)
            local_capabilities.update(dynamic_capabilities)

    if os.path.exists("./.capabilities.override.json"):
        with open("./.capabilities.override.json") as f:
            local_capabilities.update(json.load(f))

    print("capabilities:", local_capabilities)
except Exception as e:
    print("Ignoring capabilities error:")
    print(e)

capabilities = {"os": sys.platform, "arch": platform.machine(), **local_capabilities}

response = check_for_errors(
    requests.post(
        SHEETBOT_BASEURL + "/tasks/get",
        headers={**headers, "Content-Type": "application/json"},
        json={"type": "python", "capabilities": capabilities},
    )
)

json_data = response.json()
# print(json_data)
if "script" in json_data:
    os.environ["SHEETBOT_TASK_ID"] = json_data["id"]
    os.environ["SHEETBOT_AUTHORIZATION_HEADER"] = headers.get("Authorization", "")
    os.environ["SHEETBOT_TASK_BASEURL"] = SHEETBOT_BASEURL + "/tasks/" + json_data["id"]
    os.environ["SHEETBOT_TASK_ACCEPTURL"] = (
        SHEETBOT_BASEURL + "/tasks/" + json_data["id"] + "/accept"
    )
    os.environ["SHEETBOT_TASK_COMPLETEURL"] = (
        SHEETBOT_BASEURL + "/tasks/" + json_data["id"] + "/complete"
    )
    os.environ["SHEETBOT_TASK_FAILEDURL"] = (
        SHEETBOT_BASEURL + "/tasks/" + json_data["id"] + "/failed"
    )
    os.environ["SHEETBOT_TASK_DATAURL"] = (
        SHEETBOT_BASEURL + "/tasks/" + json_data["id"] + "/data"
    )
    os.environ["SHEETBOT_TASK_ARTEFACTURL"] = (
        SHEETBOT_BASEURL + "/tasks/" + json_data["id"] + "/artefacts"
    )

    check_for_errors(
        requests.post(
            os.environ["SHEETBOT_TASK_ACCEPTURL"],
            headers={**headers, "Content-Type": "application/json"},
            json={},
        )
    )

    try:
        script_response = requests.get(json_data["script"])
        script_code = script_response.text
        if "/// script" in script_code:
            with tempfile.NamedTemporaryFile(mode="w", suffix=".py", delete=False) as f:
                f.write(script_code)
                temp_file = f.name
            result = subprocess.run(
                ["uv", "run", "--script", temp_file],
                capture_output=True,
                text=True,
                env=os.environ.copy(),
            )
            os.unlink(temp_file)
            if result.returncode != 0:
                raise Exception(result.stderr)
            data = {}
        else:
            exec_globals = {}
            exec(script_code, exec_globals)
            data = exec_globals.get("data", {})
        check_for_errors(
            requests.post(
                os.environ["SHEETBOT_TASK_COMPLETEURL"],
                headers={**headers, "Content-Type": "application/json"},
                json={"data": data},
            )
        )
    except Exception as e:
        print(e)
        check_for_errors(
            requests.post(
                os.environ["SHEETBOT_TASK_FAILEDURL"],
                headers={**headers, "Content-Type": "application/json"},
                json={},
            )
        )
